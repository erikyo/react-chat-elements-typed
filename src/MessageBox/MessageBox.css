.reverse-icon-horizontal {
    transform: scaleX(-1);
}

.rce-container-mbox {
    flex-direction: column;
    display: block;
}

.rce-mbox-action {
    border-color: rgb(15 15 20 / 100%);
}

.rce-mbox-forward,
.rce-mbox-remove {
    transition: all 0.3s ease;
}

.rce-mbox-forward-left,
.rce-mbox-forward-right,
.rce-mbox-reply-btn-left,
.rce-mbox-reply-btn-right {
    opacity: 0;
    visibility: hidden;
}

.rce-container-mbox:hover .rce-mbox-forward-left,
.rce-container-mbox:hover .rce-mbox-forward-right {
    opacity: 1;
    visibility: visible;
}

.rce-mbox-remove-left,
.rce-mbox-remove-right {
    opacity: 0;
    visibility: hidden;
}

.rce-container-mbox:hover .rce-mbox-remove-left ,
.rce-container-mbox:hover .rce-mbox-remove-right,
.rce-container-mbox:hover .rce-mbox-reply-btn-left,
.rce-container-mbox:hover .rce-mbox-reply-btn-right {
    opacity: 1;
    visibility: visible;
}

.rce-mbox {
    min-width: var(--rce-mbox-min-width);
    max-width: var(--rce-mbox-max-width);
    position: relative;
    flex-direction: column;
    margin: 5px 5px 5px 20px;
    float: left;
}

.rce-mbox-body {
    box-shadow: var(--rce-mbox-filter);
    font-size: var(--rce-mbox-font-size);
    margin: 0;
    padding: var(--rce-mbox-inner-padding);
    position: relative;
    z-index: 1;
}

.rce-mbox.rce-mbox-left .rce-mbox-body {
    background-color: var(--rce-mbox-background);
    border-radius: 0 var(--rce-mbox-border-radius) var(--rce-mbox-border-radius);
    z-index: 3;
}

.rce-mbox.rce-mbox-right .rce-mbox-body {
    background-color: var(--rce-mbox-background-right);
    border-radius: var(--rce-mbox-border-radius) 0 var(--rce-mbox-border-radius) var(--rce-mbox-border-radius);
}

.rce-mbox-left .rce-mbox-body.message-focus {
    animation-name: message-box-default-focus;
    animation-duration: 1s;
}

.rce-mbox-right .rce-mbox-body.message-focus {
    animation-name: message-box-right-focus;
    animation-duration: 1s;
}

.rce-mbox.rce-mbox-right {
    float: right;
    margin-left: 5px;
    margin-right: 20px;
    border-top-right-radius: 0;
    border-top-left-radius: var(--rce-mbox-border-radius);
}

.rce-mbox-action-buttons {
    position: absolute;
    display: flex;
    top: calc(50% - 10px);
    left: calc(100% + 24px);
    flex-direction: row;
    justify-content: flex-start;
    align-items: center;
    gap: 0.5rem;
}

.rce-mbox.rce-mbox-right .rce-mbox-action-buttons {
    left: auto;
    right: calc(100% + 24px);
}

.rce-mbox-text {
    margin: 6px 0;
    word-break: break-word;
}

.rce-mbox-text-retracted {
    font-style: italic;
    user-select: none;
    display: flex;
    align-items: center;
}

.rce-mbox-text.rce-mbox-text-retracted svg {
    margin-right: 3px;
}

.rce-mbox-text-retracted.left {
    color: var(--rce-color-gray) !important;
}

.rce-mbox-text-retracted.right {
    color: var(--rce-color-gray) !important;
}

.rce-mbox-time {
    display: flex;
    justify-content: flex-end;
    gap: 4px;
    align-items: center;
    text-align: right;
    color: var(--rce-color-gray);
    font-size: 12px;
    margin-top: 10px;
    user-select: none;
}

.rce-reactions {
    position: absolute;
    bottom: 0;
    display: flex;
    gap: 4px;
    transform: translateY(50%);
}

.rce-reaction {
    text-align: center;
    background: var(--rce-color-white);
    box-shadow: 0 1px 0 rgb(0 0 0 / 7%),0 0 3px rgb(0 0 0 / 4%);
    border: 1px solid rgb(0 0 0 / 5%);
    padding: 4px;
    box-sizing: border-box;
    line-height: 20px;
    width: 30px;
    border-radius: 50%;
    cursor: pointer;
}

.rce-mbox-time.non-copiable::before {
    content: attr(data-text);
}

.rce-mbox-time-block {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    margin-right: -6px;
    margin-left: -6px;
    padding-top: 5px;
    padding-right: 3px;
    padding-bottom: 2px;
    border-bottom-left-radius: var(--rce-mbox-border-radius);
    border-bottom-right-radius: var(--rce-mbox-border-radius);
}

.rce-mbox--clear-padding {
    padding-bottom: 3px;
}

.rce-mbox.rce-mbox--clear-notch {
    border-radius: var(--rce-mbox-border-radius) !important;
}

.rce-mbox-right-notch {
    position: absolute;
    z-index: 2;
    right: -14px;
    top: 0;
    width: 15px;
    height: 15px;
    filter: drop-shadow(var(--rce-mbox-filter));
}

.rce-mbox-right-notch.message-focus {
    animation-name: message-right-notch-focus;
    animation-duration: 1s;
}

.rce-mbox-left-notch {
    position: absolute;
    z-index: 2;
    left: -14px;
    top: 0;
    width: 15px;
    height: 15px;
    filter: drop-shadow(var(--rce-mbox-filter))
}

.rce-mbox-left-notch.message-focus {
    animation-name: message-left-notch-focus;
    animation-duration: 1s;
}

.rce-mbox-title {
    margin: 0 0 8px;
    font-weight: 500;
    font-size: 13px;
    color: var(--rce-color-secondary);
    user-select: none;
    cursor: pointer;
    display: flex;
    align-items: center;
}

.rce-mbox-title:hover {
    text-decoration: underline;
}

.rce-mbox-title--clear {
    margin-bottom: 6px;
}

.rce-mbox-status {
    margin-left: 3px;
    font-size: 15px;
    height: 15px;
}

.rce-mbox-title > .rce-avatar-container {
    margin-right: 5px;
}

.rce-mbox-forwarded-message {
    position: relative;
    overflow: hidden;
    display: flex;
    border-radius: var(--rce-mbox-border-radius);
    padding: 0 0 5px;
    transition: 200ms;
    user-select: none;
}

.rce-mbox-forwarded-message-inner {
    display: flex;
    gap: 3px;
    align-items: center;
    padding: 5px 0;
    font-size: 13px;
    font-style: italic;
    color: var(--rce-color-gray-200);
}

@keyframes message-box-default-focus {
    0% {
        background-color: var(--rce-mbox-background);
    }

    50% {
        background-color: var(--rce-mbox-background-focus);
    }

    100% {
        background-color: var(--rce-mbox-background);
    }
}

@keyframes message-left-notch-focus {
    0% {
        fill: var(--rce-mbox-background);
    }

    50% {
        fill: var(--rce-mbox-background-focus);
    }

    100% {
        fill: var(--rce-mbox-background);
    }
}

@keyframes message-box-right-focus {
    0% {
        background-color: var(--rce-mbox-background-right);
    }

    50% {
        background-color: var(--rce-mbox-background-right-focus);
    }

    100% {
        background-color: var(--rce-mbox-background-right);
    }
}

@keyframes message-right-notch-focus {
    0% {
        fill: var(--rce-mbox-background-right);
    }

    50% {
        fill: var(--rce-mbox-background-right-focus);
    }

    100% {
        fill: var(--rce-mbox-background-right);
    }
}
