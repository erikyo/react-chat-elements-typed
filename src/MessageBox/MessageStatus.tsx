import type { StatusType } from "../types";
import { MdAccessTime, MdCheck, MdDoneAll } from "react-icons/md";
import type React from "react";

export function MessageStatus({ status }: { status: StatusType }) {
	return (
		<span className="rce-mbox-status">
			{status === "waiting" && (
				<MdAccessTime
					color={"var(--rce-color-gray)"}
					style={{ fontSize: 18 }}
				/>
			)}

			{status === "sent" && (
				<MdCheck
					color={"var(--rce-color-light-blue)"}
					style={{ fontSize: 18 }}
				/>
			)}

			{status === "received" && (
				<MdDoneAll color={"var(--rce-color-green)"} style={{ fontSize: 18 }} />
			)}

			{status === "read" && (
				<MdDoneAll
					color={"var(--rce-color-light-blue)"}
					style={{ fontSize: 18 }}
				/>
			)}
		</span>
	);
}
