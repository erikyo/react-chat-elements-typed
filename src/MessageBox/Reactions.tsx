import type { MessageReaction } from "../types";
import type React from "react";

export function Reactions(props: {
	reactions: MessageReaction[];
}) {
	if (!props.reactions) {
		return null;
	}
	return (
		<div className={"rce-reactions"}>
			{props.reactions?.map((reaction) => (
				<div
					className={"rce-reaction"}
					title={`${reaction.author} reacted with ${reaction.emoji}`}
					key={reaction.emoji}
				>
					{reaction.emoji}
				</div>
			))}
		</div>
	);
}
