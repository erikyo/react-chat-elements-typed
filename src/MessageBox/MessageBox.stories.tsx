import type { StoryObj } from "@storybook/react";

import MessageBox from "./MessageBox";
import type { MessageBoxType } from "../types";
import type { FC } from "react";
import { getAvatar, photo } from "../stories/example/utils/common";

const meta = {
	component: MessageBox satisfies FC<MessageBoxType>,
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default = {
	args: {
		type: "text",
		id: "1",
		title: "Title",
		titleColor: "rgba(0, 48, 255, 1)",
		text: "Text",
		content: null,
		status: "read",
		position: "left",
		focus: false,
		date: +new Date(),
		dateString: "now",
		avatar: getAvatar("avatar1"),
		retracted: false,
		forwarded: false,
		replyButton: false,
		removeButton: false,
		statusTitle: "Status",
		notch: true,
		copiableDate: true,
		className: "custom-classname",
		letterItem: null,
		reply: null,
		actionButtons: null,
	},
};

export const Simple = {
	args: {
		type: "text",
		id: "10",
		text: "Text",
	},
};

export const Reply = {
	args: {
		type: "text",
		id: "1",
		title: "Title",
		titleColor: "red",
		content: "Content",
		status: "waiting",
		position: "left",
		text: "Text",
		reply: {
			src: photo(),
			title: "Reply title",
			message: "Reply text",
			titleColor: "blue",
		},
		message: "Message text",
		replyButton: true,
	},
};

export const Forwarded = {
	args: {
		type: "text",
		id: "1",
		title: "Title",
		forwardedMessageText: "Forwarded",
		titleColor: "red",
		text: "Text",
		content: "Content",
		status: "read",
	},
};

export const Retracted = {
	args: {
		type: "text",
		id: "1",
		retracted: true,
		text: "Text",
		content: "Content",
		status: "read",
	},
};
