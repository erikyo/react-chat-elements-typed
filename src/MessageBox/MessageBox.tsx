import type { FC } from "react";
import type React from "react";
import { useEffect, useRef } from "react";
import "./MessageBox.css";

import PhotoMessage from "../PhotoMessage/PhotoMessage";
import FileMessage from "../FileMessage/FileMessage";
import SystemMessage from "../SystemMessage/SystemMessage";
import LocationMessage from "../LocationMessage/LocationMessage";
import ReplyMessage from "../ReplyMessage/ReplyMessage";
import MeetingMessage from "../MeetingMessage/MeetingMessage";
import VideoMessage from "../VideoMessage/VideoMessage";
import AudioMessage from "../AudioMessage/AudioMessage";
import MeetingLink from "../MeetingLink/MeetingLink";

import Avatar from "../Avatar/Avatar";

import { RiShareForwardFill } from "react-icons/ri";
import {
	MdAccessTime,
	MdCheck,
	MdDelete,
	MdDoneAll,
	MdMessage,
	MdOutlineEdit,
} from "react-icons/md";
import { TiArrowForward } from "react-icons/ti";

import { format } from "date-fns";

import classNames from "classnames";
import type { MessageBoxType } from "../types";
import { LeftNotch } from "../SvgIcon/leftNotch";
import { RightNotch } from "../SvgIcon/rightNotch";

const MessageBox: FC<MessageBoxType> = ({
	focus = false,
	notch = true,
	style,
	actionButtons,
	notchStyle = { color: "var(--rce-color-white)" },
	...props
}) => {
	const prevProps = useRef(focus);
	const messageRef = useRef<HTMLDivElement | null>(null);

	const positionCls = classNames("rce-mbox", {
		"rce-mbox-right": props.position === "right",
	});
	const thatAbsoluteTime =
		!/(text|video|file|meeting|audio)/g.test(props.type || "text") &&
		!(props.type === "location" && props.text);
	const dateText =
		props.date && (props.dateString || format(props.date, "HH:mm"));

	useEffect(() => {
		if (prevProps.current !== focus && focus) {
			if (messageRef?.current) {
				(messageRef.current as HTMLDivElement)?.scrollIntoView({
					behavior: "smooth",
					block: "nearest",
					inline: "start",
				});

				if (props.onMessageFocused instanceof Function)
					props.onMessageFocused(true);
			}
		}
		prevProps.current = focus;
	}, [focus, props.onMessageFocused]);

	return (
		<div
			ref={messageRef}
			className={classNames("rce-container-mbox", props.className)}
			onClick={props.onClick}
			onKeyDown={console.log}
		>
			{props.renderAddCmp instanceof Function
				? props.renderAddCmp()
				: props.renderAddCmp}
			{props.type === "system" ? (
				<SystemMessage {...props} focus={focus} notch={notch} />
			) : (
				<div
					style={style}
					className={classNames(
						positionCls,
						{ "rce-mbox--clear-padding": thatAbsoluteTime },
						{ "rce-mbox--clear-notch": !notch },
						{ "message-focus": focus },
					)}
				>
					{notch &&
						(props.position === "right" ? (
							<RightNotch style={notchStyle} focus={focus} />
						) : (
							<LeftNotch style={notchStyle} focus={focus} />
						))}
					<div className="rce-mbox-body" onContextMenu={props.onContextMenu}>
						{!props.retracted && props.forwarded === true && (
							<div
								className={classNames(
									"rce-mbox-forward",
									{ "rce-mbox-forward-right": props.position === "left" },
									{ "rce-mbox-forward-left": props.position === "right" },
								)}
								onClick={props.onForwardClick}
								onKeyDown={console.log}
							>
								<RiShareForwardFill />
							</div>
						)}

						{!props.retracted && props.replyButton && (
							<div
								className={
									!props.forwarded
										? classNames(
												"rce-mbox-forward",
												{ "rce-mbox-forward-right": props.position === "left" },
												{ "rce-mbox-forward-left": props.position === "right" },
											)
										: classNames(
												"rce-mbox-forward",
												{
													"rce-mbox-reply-btn-right": props.position === "left",
												},
												{
													"rce-mbox-reply-btn-left": props.position === "right",
												},
											)
								}
								onClick={props.onReplyClick}
								onKeyDown={console.log}
							>
								<MdMessage />
							</div>
						)}

						{!props.retracted && props.removeButton && (
							<div
								className={
									props.forwarded
										? classNames(
												"rce-mbox-remove",
												{ "rce-mbox-remove-right": props.position === "left" },
												{ "rce-mbox-remove-left": props.position === "right" },
											)
										: classNames(
												"rce-mbox-forward",
												{
													"rce-mbox-reply-btn-right": props.position === "left",
												},
												{
													"rce-mbox-reply-btn-left": props.position === "right",
												},
											)
								}
								onClick={props.onRemoveMessageClick}
								onKeyDown={console.log}
							>
								<MdDelete />
							</div>
						)}

						{(props.title || props.avatar) && (
							<div
								style={{ ...(props.titleColor && { color: props.titleColor }) }}
								onClick={props.onTitleClick}
								onKeyDown={console.log}
								className={classNames("rce-mbox-title", {
									"rce-mbox-title--clear": props.type === "text",
								})}
							>
								{props.avatar && (
									<Avatar
										size={"xsmall"}
										letterItem={props.letterItem}
										src={props.avatar}
									/>
								)}
								{props.title && !props.reply && <span>{props.title}</span>}
							</div>
						)}

						{props.forwardedMessageText ? (
							<div className="rce-mbox-forwardedMessage">
								<div className="rce-mbox-forwarded-message">
									<TiArrowForward fontSize={18} />
									<i style={{ margin: "0 3px 1px 0" }}>
										{" "}
										{props.forwardedMessageText}
									</i>
								</div>
							</div>
						) : null}

						{!props.forwardedMessageText && props.reply ? (
							<ReplyMessage
								onClick={() => props.onReplyMessageClick}
								{...props.reply}
							/>
						) : null}

						{props.type === "text" && (
							<div
								className={classNames("rce-mbox-text", {
									"rce-mbox-text-retracted": props.retracted,
									left: props.position === "left",
									right: props.position === "right",
								})}
							>
								{!props.forwardedMessageText && props.retracted && (
									<MdOutlineEdit />
								)}
								{props.text}
							</div>
						)}

						{props.type === "location" && (
							<LocationMessage focus={focus} notch={notch} {...props} />
						)}

						{props.type === "photo" && (
							<PhotoMessage focus={focus} notch={notch} {...props} />
						)}

						{props.type === "video" && (
							<VideoMessage focus={focus} notch={notch} {...props} />
						)}

						{props.type === "file" && (
							<FileMessage focus={focus} notch={notch} {...props} />
						)}

						{props.type === "meeting" && (
							<MeetingMessage focus={focus} notch={notch} {...props} />
						)}
						{props.type === "audio" && (
							<AudioMessage focus={focus} notch={notch} {...props} />
						)}

						{props.type === "meetingLink" && (
							<MeetingLink {...props} actionButtons={actionButtons} />
						)}

						<div
							title={props.statusTitle}
							className={classNames(
								"rce-mbox-time",
								{ "rce-mbox-time-block": thatAbsoluteTime },
								{ "non-copiable": !props.copiableDate },
							)}
							data-text={props.copiableDate ? undefined : dateText}
						>
							{props.copiableDate &&
								props.date &&
								(props.dateString || format(props.date, "HH:mm"))}
							{props.status && (
								<span className="rce-mbox-status">
									{props.status === "waiting" && (
										<MdAccessTime
											color={"var(--rce-color-gray)"}
											style={{ fontSize: 18 }}
										/>
									)}

									{props.status === "sent" && (
										<MdCheck
											color={"var(--rce-color-light-blue)"}
											style={{ fontSize: 18 }}
										/>
									)}

									{props.status === "received" && (
										<MdDoneAll
											color={"var(--rce-color-green)"}
											style={{ fontSize: 18 }}
										/>
									)}

									{props.status === "read" && (
										<MdDoneAll
											color={"var(--rce-color-light-blue)"}
											style={{ fontSize: 18 }}
										/>
									)}
								</span>
							)}
						</div>
					</div>
				</div>
			)}
		</div>
	);
};

export default MessageBox;
