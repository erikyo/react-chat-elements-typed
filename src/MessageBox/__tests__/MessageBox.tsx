import { describe, expect, it } from "vitest";
import React from "react";
import MessageBox from "../MessageBox";
import { render } from "@testing-library/react";

describe("Avatar component", () => {
	it("should render without issues", () => {
		const component = render(<MessageBox status="read" />);
		expect(component.container).toBeTruthy();
		expect(component.container).toMatchSnapshot();
	});
});
