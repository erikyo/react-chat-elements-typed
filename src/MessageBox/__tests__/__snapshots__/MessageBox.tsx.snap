// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Avatar component > should render without issues 1`] = `
<div>
  <div
    class="rce-container-mbox"
  >
    <div
      class="rce-mbox rce-mbox-left"
    >
      <svg
        class="rce-mbox-left-notch"
        height="20"
        role="img"
        style="display: inline-flex; vertical-align: middle; fill: var(--rce-mbox-background);"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 0v20L0 0"
        />
      </svg>
      <div
        class="rce-mbox-body"
      >
        <div
          class="rce-mbox-action-buttons"
        >
          <button
            class="rce-button rce-mbox-action rce-mbox-forward circle"
            style="background-color: rgba(110, 110, 120, 0.5); border-color: rgba(110, 110, 120, .5); border-width: 0px; flex-direction: row; padding: 6px;"
            type="button"
          >
            <span
              class="rce-button-icon--container"
            >
              <span
                class="rce-button-icon"
                style="float: inherit; font-size: 12px;"
              >
                <svg
                  class="reverse-icon-horizontal"
                  fill="currentColor"
                  height="20"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M7 8V5l-7 7 7 7v-3l-4-4 4-4zm6 1V5l-7 7 7 7v-4.1c5 0 8.5 1.6 11 5.1-1-5-4-10-11-11z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
        <div
          class="rce-mbox-time non-copiable"
        >
          <span
            class="rce-mbox-status"
          >
            <svg
              color="var(--rce-color-light-blue)"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="font-size: 18px;"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m18 7-1.41-1.41-6.34 6.34 1.41 1.41L18 7zm4.24-1.41L11.66 16.17 7.48 12l-1.41 1.41L11.66 19l12-12-1.42-1.41zM.41 13.41 6 19l1.41-1.41L1.83 12 .41 13.41z"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
